src/main/java/com/google/sps/DataServlet.java

```diff
@@ -179,118 +193,162 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
       Set<Integer> mutationIndicesSet = new HashSet<>();
 
+      // Add all mutations relevant to on-screen nodes
       mutationIndicesSet.addAll(
           findRelevantMutations(truncatedGraphNodeNamesNext, mutationIndicesMap, mutList));
-      mutationIndicesSet.addAll(getMutationIndicesOfToken(tokenParam, mutList));
+
+      // Add all mutations relevant to the queried token, computing and caching it if it
+      // hasn't been done already
+      if (!tokenIndicesMap.containsKey(tokenNameParam)) {
+        tokenIndicesMap.put(tokenNameParam, getMutationIndicesOfToken(tokenNameParam, mutList));
+      }
+      mutationIndicesSet.addAll(tokenIndicesMap.get(tokenNameParam));
+
+      // Add all mutations relevant to the queried node names
       mutationIndicesSet.addAll(findRelevantMutations(nodeNames, mutationIndicesMap, mutList));
       filteredMutationIndices = new ArrayList<>(mutationIndicesSet);
       Collections.sort(filteredMutationIndices);
 
-      // Show mutations relevant to nodes that used to have the token but
-      // might not exist anymore and the queried nodes
+      // Show mutations relevant to nodes that are related to on-screen nodes, nodes that
+      // used to/still have the queried token and any queried nodes
       filteredDiff = filterMultiMutationByNodes(diff, Sets.union(truncatedGraphNodeNames, queried));
     }
+
+    /*
+     ***********************
+     * Error Handling
+     ***********************
+     */
+
     // We set the headers in the following 4 scenarios:
-    // The searched node is not in the graph and is never mutated
+    // truncatedGraph.nodes().size() == 0 means something was queried but wasn't found
+    // in the graph
+    // filteredMutationIndices.size() == 0 means the searched object is never mutated
     if (truncatedGraph.nodes().size() == 0 && filteredMutationIndices.size() == 0) {
       response.setHeader(
-          "serverError", "The searched node does not exist anywhere in this graph or in mutations");
+          "serverError",
+          "The searched node/token does not exist anywhere in this graph or in mutations");
       return;
     }
-    // The searched node is not in the graph but is mutated at some past/future
-    // point. The diff conditions are included to prevent entry into this case
-    // when the searched node is deleted for example. The diff being non-empty
-    // means that there is some mutation pertaining to the searched node to show
+    // truncatedGraph.nodes().size() == 0 means something was queried but wasn't found
+    // in the graph
+    // filteredMutationIndices.size() != 0 means the searched object is mutated at some
+    // point
+    // filteredDiff == null || filteredDiff.getMutationList().size() == 0 means that none
+    // of the searched objects are mutated here. This exists to prevent this message from
+    // being emitted when the searched object is deleted in this graph so it doesn't exist
+    // but is still mutated
```

Comment:
nit: same here, but you might need to shorten the comments to fit within one line, something like:
```
if (truncatedGraph.nodes().size() == 0 // searched object not found in the graph
        && filteredMutationIndices.size() != 0 // ... and mutated at some point
        && (filteredDiff == null || filteredDiff.getMutationList().size() == 0)) { // ... but not mutated here
```